src/modm/driver/inertial/mmc5603.hpp

```diff
@@ -0,0 +1,342 @@
+		/// non-zero value and a 1 has been written into Cmm_freq_en. The
+		/// internal counter will start counting as well since this bit is set.
+		Cmm_en = Bit4,
+		/// Factory use only, reset value is 0.
+		INT_mdt_en = Bit5,
+		/// Factory use only, reset value is 0.
+		INT_meas_done_en = Bit6,
+		/// If this bit is set to 1 to achieve 1000Hz ODR.
+		Hpower = Bit7,
+	};
+	MODM_FLAGS8(Control2);
+
+	/// These bits determine how many measurements are done before a set is
+	/// executed, when the part is in continuous mode and the automatic
+	/// set/reset is enabled. From 000 to 111, the sensor will do one set
+	/// for every 1, 25, 75, 100, 250, 500, 1000, and 2000 samples. In order
+	/// to enable this feature, both En_prd_set and Auto_SR must be set to
+	/// 1, and the part should work in continuous mode. Please note that
+	/// during this operation, the sensor will not be reset.
+	enum class
+	PeriodicalSet : uint8_t
+	{
+		Samples1 = 0x00,
+		Samples25 = 0x01,
+		Samples75 = 0x02,
+		Samples100 = 0x03,
+		Samples250 = 0x04,
+		Samples500 = 0x05,
+		Samples1000 = 0x06,
+		Samples2000 = 0x07,
+	};
+	MODM_FLAGS_CONFIG(Control2, PeriodicalSet);
+
+	struct modm_packed
+	Data
+	{
+		template < class I2cMaster >
+		friend class Mmc5603;
+
+		// DATA ACCESS
+		/// returns the magnetic field in Gauss
+		///@{
+		float inline
+		x() const { return swapData(0); }
+
+		float inline
+		y() const { return swapData(1); }
+
+		float inline
+		z() const { return swapData(2); }
+		///@}
+
+		/// returns the temperature in Celcius
+		int8_t inline
+		t() const { return int16_t(data[9]) * 4/5 - 75; }
+
+		float inline
+		operator [](uint8_t index) const
+		{ return (index < 3) ? swapData(index) : 0; }
+
+	private:
+		uint8_t data[10];
+
+		float inline
+		swapData(uint8_t index) const
+		{ return (int32_t((data[2*index] << 8) | data[2*index+1]) - 32768) / 1024.f; }
+	};
+
+protected:
+	/// @cond
+	static constexpr uint8_t
+	i(Register reg) { return uint8_t(reg); }
+	static constexpr uint8_t
+	i(Bandwidth bw) { return uint8_t(bw); }
+	static constexpr uint8_t
+	i(PeriodicalSet ps) { return uint8_t(ps); }
+	using Register_t = FlagsGroup<Control0_t, Control1_t, Control2_t>;
+	/// @endcond
+}; // struct mmc5603
+
+/**
+ * @ingroup modm_driver_mmc5603
+ * @author	Niklas Hauser
+ */
+template < class I2cMaster >
+class Mmc5603 : public mmc5603, public modm::I2cDevice< I2cMaster, 3 >
+{
+public:
+	/// Constructor, requires a mmc5603::Data object
+	Mmc5603(Data &data, uint8_t address=addr())
+	:	I2cDevice<I2cMaster,3>(address), data(data)
+	{}
+
+	modm::ResumableResult<bool>
+	startTemperatureMeasurement()
+	{ return update(Register::InternalControl0, Control0::Take_meas_T); }
+
+	modm::ResumableResult<bool>
+	configureContinuousMode(uint8_t frequency)
+	{
+		RF_BEGIN();
+		if (not RF_CALL(write(Register::ODR, frequency)))
+			RF_RETURN(false);
+		if (not RF_CALL(update(Register::InternalControl1, Register_t(i(Bandwidth::Ms1_2)))))
+		    RF_RETURN(false);
```

Comment:
Wouldn't it make sense to be able to configure the bandwidth if the measurement frequency is configurable?

```suggestion
	modm::ResumableResult<bool>
	configureContinuousMode(uint8_t frequency, Bandwidth bandwidth = Bandwidth::Ms1_2)
	{
		RF_BEGIN();
		if (not RF_CALL(write(Register::ODR, frequency)))
			RF_RETURN(false);
		if (not RF_CALL(update(Register::InternalControl1, Register_t(i(bandwidth)))))
		    RF_RETURN(false);
```

Reply:
Yes, also the 1.2ms measurement time was only necessary for 1000Hz operation, however, that does not work with Automatic Set/Reset as enabled in the function, hence the frequency only being 8-bit for a max of 255Hz. If you need 1kHz update rate, then you need to do it manually or update the driver later.
![](https://user-images.githubusercontent.com/163066/107772629-c2499680-6d3c-11eb-95a2-9a0f7b3d06ab.png)